composer/trainer/mosaic_fsdp_utils.py

```diff
@@ -0,0 +1,894 @@
+    if auto_wrap_policy is not None:
+        auto_wrap_kwargs = {
+            'module': module,
+            'auto_wrap_policy': auto_wrap_policy,
+            'wrapper_cls': FullyShardedDataParallel,
+            'ignored_modules': self._ignored_modules,
+            'ignored_params': self._ignored_params,
+            'only_wrap_children': True,  # avoid double wrapping the root
+        }
+        fsdp_kwargs = {
+            'process_group': process_group,
+            'sharding_strategy': sharding_strategy,
+            'cpu_offload': cpu_offload,
+            'backward_prefetch': backward_prefetch,
+            'mixed_precision': mixed_precision,
+            'param_init_fn': param_init_fn,
+            'device_id': device_id,
+            'sync_module_states': sync_module_states,
+            'forward_prefetch': forward_prefetch,
+            'limit_all_gathers': limit_all_gathers,
+            'use_orig_params': use_orig_params,
+        }
+        if sharding_strategy in HYBRID_SHARDING_STRATEGIES:
+            # Share root process groups with children to maintain
+            # the invariant that all FSDP modules will have the same
+            # process groups.
+            fsdp_kwargs['process_group'] = (self.process_group, self._inter_node_pg)
+
+        # call the custom _auto_wrap function
+        _custom_auto_wrap_t2p0p1(
+            auto_wrap_kwargs, fsdp_kwargs, FullyShardedDataParallel
+        )
+
+    backward_prefetch_limit = 1
+    forward_prefetch_limit = 1
+    _init_core_state(
+        self,
+        sharding_strategy,
+        mixed_precision,
+        cpu_offload,
+        limit_all_gathers,
+        use_orig_params,
+        backward_prefetch_limit,
+        forward_prefetch_limit,
+    )
+    _init_runtime_state(self)
+    _init_prefetching_state(self, backward_prefetch, forward_prefetch)  # type: ignore
+    _init_buffer_state(self, module)
+    _init_param_handle_from_module(
+        self,
+        module,
+        device_id,
+        param_init_fn,
+        sync_module_states,
+        FullyShardedDataParallel,
+    )
+    self._fsdp_wrapped_module = module
+    if not use_orig_params:
+        _check_orig_params_flattened(self, self._ignored_params)
+        _register_flat_param(self, self)
+
+    # `_state_dict_type` controls the `state_dict()` behavior, which is
+    # implemented using post-save and pre-load hooks
+    _init_state_dict_state(self)
+    _register_all_state_dict_hooks(self)
+
+
+def get_split_size(dim_size: int, chunks: int) -> int:
+    """Gets the minimum size per chunk.
+
+    A tensor of dim_size 5 and 4 chunks will have chunks of size [2, 1, 1, 1].
+
+    Args:
+        dim_size(int): Size of the dimension being chunked.
+        chunks(int): Number of chunks to create for ``dim_size``.
+
+    Returns:
+        An int indicating the split size to use.
+    """
+    return dim_size // chunks
+
+
+def get_chunked_dim_size(dim_size: int, chunks: int, idx: int) -> int:
```

Comment:
can we also tag all the fns for chunking similarly to fsdp wrap with torch nightly release candidate 3?

Reply:
Using this for both 2.0.1 and 2.1.0 so leaving version tag off